### ISSUE TITLE
Make Applewin 'portable'.

### ISSUE THREAD

**Issue by Keatah** [AppleWin/AppleWin#446]
Is there a way to make AppleWin behave like a 'portable' app and have it save its config to an ini file instead of the registry? And is there a way to disable the "I accept" & "I decline" dialog - which always pops up when running a different version..?

When doing curating and archiving activities I often switch back and forth between 2 versions. 1 for better video modes and speed, the other for better compatibility.

**Comment by TommyGH** [comment]
"And is there a way to disable the "I accept" & "I decline" dialog - which always pops up when running a different version..?"

Interesting, I just made a comment about this "I accept" and "I decline" in another issue.  Until now I was unaware that it was due to my switching back and forth between versions!  And I second the motion to disable the pop up (if possible).

**Comment by tomcw** [comment]
>And is there a way to disable the "I accept" & "I decline" dialog - which always pops up when running a different version..?

This is the prompt showing the GPL license. I guess we could change from:
- show prompt if version changes (since the last accepted version)

To
- show prompt if version is newer than the last accepted version

NB. If you want to suppress this, then you'd need to set the correct `Version` in the Registry before starting that version of AppleWin:
HKEY_CURRENT_USER\Software\AppleWin\CurrentVersion\Configuration\Version

**Comment by tomcw** [comment]
>Is there a way to make AppleWin behave like a 'portable' app and have it save its config to an ini file instead of the registry?

Not currently.

I guess you could write a script to export the sub-branch:
HKEY_CURRENT_USER\Software\AppleWin\CurrentVersion\Configuration
Then import the other version's sub-branch.

This would also "solve" the above issue with the prompt about accepting the GPL each time.

**Comment by tomcw** [comment]
But really you want #357 done, which is the underlying problem.

**Comment by Keatah** [comment]
Yes I primarily switch because of video modes. The second reason is speed. the older the version the faster it runs. more or less I suppose.

To eliminate the dialog I think I'll make a tiny batch file that plugs the version # I'm using into the registry.

**Comment by tomcw** [comment]
>The second reason is speed. the older the version the faster it runs. more or less I suppose.

You must be talking about the full-speed mode. But specifically is this during **regular** emulation or debugger stepping emulation? If it's the latter, then there is a `gg` command to "Go (unthrottled)" in single-stepping mode (eg. with breakpoints active). This was added at experimental v1.26.2.0.

**Comment by Keatah** [comment]
I would be referring to standard/regular emulation. Nothing in debugger.

**Comment by tomcw** [comment]
>The second reason is speed. the older the version the faster it runs. more or less I suppose.

What older version(s) are you using?

For standard/regular emulation (at full-speed) then 1.25.x and 1.26.2+ should be very comparable (skip to the end of #347 for up-to-date numbers). But if you have a specific use-case where this isn't the case, then let me know, and I'll take a look.

Of course going way back (eg. v1.12.n, from 2005) then this used x86 assembly, and was way faster than the C code we use today (but completely unmaintainable!).

**Comment by tomcw** [comment]
Oh, and don't rely on the built-in speed test - see #424.

**Comment by rzumer** [comment]
That's something I would like to see as well. If not enabled by default, it could be implemented using a preprocessor flag and distributed as an alternative build. I might give it a shot when I have more time to look into it.

Even if it involves disabling features like file type associations, remembering the working directory and loaded files between sessions, portability is more important to me. My main issue with versions conflicting with one another is when I am testing changes and using a compiled build vs. testing using a different version of AppleWin. The license prompt is displayed again and configuration within each installation does not stay fixed. The ideal setup in my opinion would be the emulator only writing/defaulting to writing within the directory where it is installed (including screenshots, save states, etc.), and keeping all its configuration there.

Regarding the GPL license prompt: is it a choice to show it on initial runs, or a legal constraint based on an interpretation of the license terms? I believe that most applications licensed under the GPL simply include a menu option that displays license information when selected.

**Comment by tomcw** [comment]
>[...] and distributed as an alternative build
>I might give it a shot when I have more time to look into it.

Alt builds are not a good idea.
Ideally go the yaml route, similar to save-states - create a PR then we can look at merging into the main-line.

>Regarding the GPL license prompt: is it a choice to show it on initial runs, or a legal constraint based on an interpretation of the license terms?

Legally it can just be a text file alongside the exe (which AppleWin has too). But I felt this was too easy for someone (malicious) to slice off, and then there's nothing in the AppleWin application that mentions GPL (ie. there's no Help->About message box... and in AppleWin there's no application menu bar for simplified UI).

I chose the current "show on initial run" as this is how Notepad++ does it, and at the time that seemed like a good fit.

**Comment by sicklittlemonkey** [comment]
The GPL notice would be nicer if only shown once, not for every new version.

Perhaps for a "portable" version there could be a -agree-gpl command line
argument.

(Said without looking at the code, apologies.)

Cheers,
Nick.

**Comment by tomcw** [comment]
NB. PR #757 adds support for `applewin -conf <ini>`

I'll keep this open, as it doesn't solve the suppression of the GPL notice.

**Comment by Keatah** [comment]
Why not move that functionality to the .ini file? Or do away with it entirely? Wouldn't the help file suffice for the notification?

**Comment by tomcw** [comment]
>Wouldn't the help file suffice for the notification?

See above ([here](https://github.com/AppleWin/AppleWin/issues/446#issuecomment-480482811)). I want the GPL to be part of the AppleWin exe so that it can't be spliced off.

Currently Nick's idea seems the best:
>Perhaps for a "portable" version there could be a -agree-gpl command line argument.

But to complement this, I'd still like some way to be able to show the GPL dialog from the AppleWin app, eg. right-click on the `Help` icon -> `About`  (which would show the GPL).

**Comment by sicklittlemonkey** [comment]
> I'd still like some way to be able to show the GPL dialog from the
AppleWin app, eg. right-click on the Help icon -> About (which would show
the GPL).

You could add an about tab in the setup and show that on first open (since
you're keeping a flag) or even have the checkbox on that screen.

Precious real estate, I know.

Cheers,
Nick.
